build/WFCORE-3769_WFLY-10305_galleon_build.adoc

```diff
@@ -0,0 +1,166 @@
+ ** This distribution should be produced from the 'dist' maven module and should
+    have the org.wildfly:wildly-dist maven groupId and artifactId. This
+    is the traditional location and maven GA for our 'thin' distribution.
+ * The default WildFly Core maven build should continue to use the legacy 
+   org.wildfly.build:wildfly-feature-pack-build-maven-plugin to produce a 'legacy
+   core feature pack'. The maven module for this should be the 'core-feature-pack'
+   module, and its maven coordinate should be the same
+   org.wildfly.core:wildfly-core-feature-pack used in previous releases.
+ * The default WildFly maven build should continue to use the legacy 
+   org.wildfly.build:wildfly-feature-pack-build-maven-plugin to produce a 'legacy
+   servlet feature pack'. The maven module for this should be the 'servlet-feature-pack'
+   module, and its maven coordinate should be the same
+   org.wildfly:wildfly-servlet-feature-pack used in previous releases.
+ * The default WildFly Core maven build should continue to use the legacy 
+   org.wildfly.build:wildfly-feature-pack-build-maven-plugin to produce a 'legacy
+   full feature pack'. The maven module for this should be the 'feature-pack'
+   module, and its maven coordinate should be the same
+   org.wildfly:wildfly-feature-pack used in previous releases.
+ * These legacy feature packs should be produced in such a way that their output
+   would be acceptable for release as the primary build output if this shift to 
+   Galleon had not occurred. In other words, these legacy feature packs should be
+   usable by users who are accustomed to consuming our feature packs.
+ * The WildFly Core maven build should include a profile that builds
+   new maven modules that produce 'legacy thin and fat distributions' using the legacy
+   org.wildfly.build:wildfly-server-provisioning-maven-plugin. These will use
+   the 'legacy core feature pack' discussed previously.
+ ** This maven profile should not run by default.
+ * The WildFly maven build should include a profile that builds
+   new maven modules that produce 'legacy servlet and full thin and fat distributions' 
+   using the legacy org.wildfly.build:wildfly-server-provisioning-maven-plugin. These will
+   use the 'legacy servlet feature pack' and 'legacy full feature pack' discussed previously.
+ ** This maven profile should not run by default.
+ * These legacy distributions should be produced in such a way that their output, except 
+   for their maven GAVs, would be acceptable for release as the primary build output if 
+   this shift to Galleon had not occurred. In other words, these distributions should 
+   provide a valid base for comparison to the output of the Galleon build.
+ ** The only purpose of these legacy distributions is to facilitate such comparisons.
+ * The content of the Galleon-produced distributions should functionally match the output
+   of the legacy distributions, with only minimal minor cosmetic differences.
```

Comment:
Unfortunately I won't guarantee identical content. We used to have comments in our config files, for example. New configs are generated using embedded controller. There is no way atm to generate comments, for example.
Another example, the undertow subsystem config element allows to configure a few attributes that if not explicitly configured assume default values. These values are not present in the current config but are initialized to their default values in the galleon generated config. In general, it is not required to initialize all the parameters but this specific case does require it because these attributes are used to resolve capabilities that are ultimately used to validate the config before the installation. We can remove those attributes but once we do that we can't guarantee the configuration produced by galleon is going to work. In all our configs this is the only place where I had to explicitly use the default values.
Other possible cosmetic diff, the order of some attributes or elements may be different. I am trying to preserve the order but in a few places that could still end being slightly different.

Reply:
Comments in our standard config files are bugs, as they result in irrelevant diffs when the file gets repersisted (since we can't store comments + their xml location in the model and thus can't persist them.) So not having them in galleon is not a flaw, it's a reason to clean up existing bugs.

The comments in our current standard configs are in the domain.xml and host[...].xml variants. We long ago cleaned them out of the standalone variants.

https://issues.jboss.org/browse/WFCORE-3771
https://issues.jboss.org/browse/WFLY-10311

Reply:
I don't believe xml *attribute* ordering in the standard config file should be different due to galleon. If there is a difference, that indicates a bug in the legacy configs or a bug in the xml persistence (i.e. not related to galleon).  We've long had a rule that our standard configs should follow xsd order, as should our xml persisters.  If a config doc created by our xml persistence code (which is what the galleon produced docs are) is in a different *attribute* order from the legacy configs, then that means either the xml persister was wrong or the legacy config was wrong. Most every release we have an exercise to find and correct those kinds of problems. So for WF 13 we should use the diff of the galleon build vs legacy build to help find and clean up those.  In core I've found and sent patches for two such issues:

https://issues.jboss.org/browse/WFCORE-3772
https://issues.jboss.org/browse/WFCORE-3773

XML element order is a different matter. XML element order typically reflects the order in which independent management resources are added, and that can vary between the legacy configs and what galleon produces. That's for valid reasons.  The legacy configs were manually hand coded as a single entire config with element order set to whatever the human author manually wrote. A galleon config is created from groups of features, with the content of those groups organized around convenient reuse and convenient exclusion of less commonly used features. Composing a config from those distinct groups of features can result in a different order of elements. 

The legacy config generator also seems to have a bug in the order in which it creates the extension elements. (This is one area of the legacy configs that is not under manual control.) Galleon does not have this bug.